### PR TITLE
GH-123516: Improve JIT memory consumption by invalidating cold executors

### PR DESCRIPTION
**Pull request by savannahostrowski**
This PR aims to improve memory usage of the JIT by periodically invalidating "cold" executors (more about preliminary results in https://github.com/python/cpython/pull/123402#issuecomment-2319188505). More specifically, this PR:
- Introduces a new field on `PyExecutorObject`'s `vm_data` called `warm`, which serves as a boolean to track whether an executor has been run (`warm` is incremented in a new uop `_MAKE_WARM`, tacked on to the beginning of the trace)
- Every 10 executor creations (set via `JIT_CLEANUP_THRESHOLD`), we run `_Py_Executors_InvalidateCold` as part of `PyOptimizer_Optimize` (where executors are created)
- `_Py_Executors_InvalidateCold` is a new function that enables us to traverse the linked list of executors (stored on `PyExecutorObject`) and invalidate any executors that have not been run (because it's not being used and it's fairly cheap to recreate if need be).

I'll also note that there's more that we could try here (I've got some other experiments that I'd like to run), but I wanted to at least start by landing these changes.

* Issue: gh-123516

### CODE REVIEW

File: Python/optimizer.c [hunk 1]
```diff
@@ -1659,4 +1668,61 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
...
+void _Py_Executor_Invalidate(_PyExecutorObject *executor)
+{
+   if (executor->vm_data.valid) {
+    unlink_executor(executor);
+    executor_clear(executor);
+   }
+}
+
+void
+_Py_Executors_InvalidateOld(PyInterpreterState *interp, int is_invalidation)
```
Comment (savannahostrowski) [inline Python/optimizer.c]: Perhaps this should be renamed to `InvalidateCold`, given the direction this ended up going? Happy to update/take suggestions

**Comment by brandtbucher** [comment]
Looking at this now!

Since we'll probably iterate on heuristics for invalidation later, do you mind creating a new issue just for cold executor invalidation, and link that one instead?

File: Include/internal/pycore_optimizer.h [hunk 2]
```diff
@@ -73,6 +73,7 @@ typedef struct _PyExecutorObject {
     uint32_t exit_count;
     uint32_t code_size;
     size_t jit_size;
+    uint32_t run_count;
```
Comment (brandtbucher) [inline Include/internal/pycore_optimizer.h]: Since this is effectively a boolean field, can you make it part of the bitfield on the `_PyVMData` struct (right after `valid` and `linked`)? You can steal a bit from `chain_depth`, since it doesn't need 14. Then we can treat the field like a boolean everywhere.

File: Misc/NEWS.d/next/Core_and_Builtins/2024-08-27-21-44-14.gh-issue-116017.ZY3yBY.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Improved JIT memory consumption by invalidating cold executors
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-08-27-21-44-14.gh-issue-116017.ZY3yBY.rst]: Since this is user-facing, I'd reword it to avoid mentioning implementation details (nobody knows what an "executor" is) and focus on the user-facing impact. For example, I'd emphasize that the improvement happens because the memory for "cold" (or infrequently-executed) code will now be freed periodically (instead of living forever). This is pretty cool, so don't be afraid to write a few sentences if it would help explain how programs will benefit, etc. :)

File: Python/bytecodes.c [hunk 4]
```diff
@@ -4844,6 +4844,10 @@ dummy_func(
             assert(((_PyExecutorObject *)executor)->vm_data.valid);
         }
 
+        tier2 op(_INCREMENT_RUN_COUNT, (--)) {
+            current_executor->run_count++;
```
Comment (brandtbucher) [inline Python/bytecodes.c]: If this becomes a bitfield, we can just store `true` here. That should be a *very* tiny bit faster!

```suggestion
            current_executor->run_count = true;
```

File: Python/optimizer.c [hunk 5]
```diff
@@ -157,6 +157,7 @@ _Py_SetTier2Optimizer(_PyOptimizerObject *optimizer)
     return old == NULL ? -1 : 0;
 }
 
+int executors_created = 0;
```
Comment (brandtbucher) [inline Python/optimizer.c]: Let's make this a `size_t`, and stick it on the `PyInterpreterState` struct, next to the linked list of executors. We avoid using globals like this since it can break in the presence of multiple interpreters or threads.

File: Python/optimizer.c [hunk 6]
```diff
@@ -182,6 +183,12 @@ _PyOptimizer_Optimize(
     if (err <= 0) {
         return err;
     }
+
+    if (++executors_created >= 10) {
```
Comment (brandtbucher) [inline Python/optimizer.c]: Maybe `#define` a useful name for `10` here, like `JIT_CLEANUP_THRESHOLD` or something. You can stick it in `pycore_optimizer.h`.

File: Python/optimizer.c [hunk 7]
```diff
@@ -1194,6 +1202,7 @@ make_executor_from_uops(_PyUOpInstruction *buffer, int length, const _PyBloomFil
     executor->jit_code = NULL;
     executor->jit_side_entry = NULL;
     executor->jit_size = 0;
+    executor->run_count = UINT32_MAX;
```
Comment (brandtbucher) [inline Python/optimizer.c]: If this becomes a boolean, we can just make this `true`.

Also, can you add a comment explaining why we set this (so the executor isn't immediately detected as cold and wiped out)?

File: Python/optimizer.c [hunk 8]
```diff
@@ -182,6 +183,12 @@ _PyOptimizer_Optimize(
     if (err <= 0) {
         return err;
     }
+
+    if (++executors_created >= 10) {
+        executors_created = 0;
+        _Py_Executors_InvalidateOld(interp, 0);
```
Comment (brandtbucher) [inline Python/optimizer.c]: I haven't thought about it much, but it might make sense to do this *before* calling `opt->optimize`. That way peak memory usage is reduced, since we free *then* allocate, rather than allocate *then* free.

File: Python/optimizer.c [hunk 9]
```diff
@@ -1659,4 +1668,61 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
     }
 }
 
+void _Py_Executor_Invalidate(_PyExecutorObject *executor)
+{
+   if (executor->vm_data.valid) {
+    unlink_executor(executor);
+    executor_clear(executor);
+   }
+}
```
Comment (brandtbucher) [inline Python/optimizer.c]: If we're not using this, let's get rid of it. It's not particularly hard to implement if/when we find the need. :)

File: Python/optimizer.c (hunk 1 above)
Comment (brandtbucher) [reply Python/optimizer.c]: You read my mind! Yeah, let's rename this.

Comment (brandtbucher) [reply Python/optimizer.c]: Also, I don't think we need the `is_invalidation` arg here. It's always called with `0`.

File: Python/optimizer.c [hunk 10]
```diff
@@ -1659,4 +1668,61 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
...
+
+void
+_Py_Executors_InvalidateOld(PyInterpreterState *interp, int is_invalidation)
+{
```
Comment (brandtbucher) [inline Python/optimizer.c:1674]: So, most of this body was copy-pasted from an earlier version of `_Py_Executors_InvalidateDependency`. It's since been cleaned up a bit... maybe re-copy-and-paste the implementation to use the simpler version? Just make sure to keep your changes.

I'm wondering if there's a way the code could be shared for both somehow, but that might be more trouble than it's worth...

File: Python/optimizer.c [hunk 11]
```diff
@@ -1659,4 +1668,61 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
...
+        exec = next;
+    }
+    if (invalidate != NULL) {
+        for (Py_ssize_t i = 0; i < PyList_GET_SIZE(invalidate); i++) {
+            _PyExecutorObject *exec = (_PyExecutorObject *)PyList_GET_ITEM(invalidate, i);
+            executor_clear(exec);
+        }
+        Py_DECREF(invalidate);
+    }
+    return;
```
Comment (brandtbucher) [inline Python/optimizer.c:1702]:
```suggestion
```

**Review by brandtbucher — COMMENTED** [review]
So excited to see this! Some thoughts I had while reading:

**Comment by brandtbucher** [comment]
Also, for reference:
- [No perf impact.](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240820-3.14.0a0-071a1a4-JIT/bm-20240820-linux-x86_64-savannahostrowski-jit_mem_invalidate_1-3.14.0a0-071a1a4-vs-base.svg)
- 1.5% (up to 8% on some benchmarks) [reduction in memory](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240820-3.14.0a0-071a1a4-JIT/bm-20240820-linux-x86_64-savannahostrowski-jit_mem_invalidate_1-3.14.0a0-071a1a4-vs-base-mem.svg) for the entire process!
- Only 2% fewer traces executed and less than 1% fewer uops executed (if you discount the new one that was added), with slightly longer average trace run length and no real impact on tier one instructions executed.

All of this shows that we're having the desired impact of reducing memory without actually wiping out much hot code.

File: Include/internal/pycore_optimizer.h (hunk 2 above)
Comment (savannahostrowski) [reply Include/internal/pycore_optimizer.h]: Ha! That's very true, since we ended up with a threshold of 1. Very clever and some nice cleanup after that experimentation

File: Python/optimizer.c [hunk 12]
```diff
@@ -1659,4 +1669,46 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
...
+        }
+
+        exec = next;
+    }
+    for (Py_ssize_t i = 0; i < PyList_GET_SIZE(invalidate); i++) {
+        _PyExecutorObject *exec = (_PyExecutorObject *)PyList_GET_ITEM(invalidate, i);
+        executor_clear(exec);
+    }
+        Py_DECREF(invalidate);
+        return;
```
Comment (sobolevn) [inline Python/optimizer.c]:
```suggestion
    Py_DECREF(invalidate);
    return;
```

Looks like they have too many spaces.

File: Python/optimizer.c [hunk 13]
```diff
@@ -1657,6 +1667,49 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
             OPT_STAT_INC(executors_invalidated);
         }
     }
+    interp->executors_created=0;
```
Comment (picnixz) [inline Python/optimizer.c]:
```suggestion
    interp->executors_created = 0;
```

File: Python/optimizer.c [hunk 14]
```diff
@@ -1657,6 +1667,49 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
...
+    for (_PyExecutorObject *exec = interp->executor_list_head; exec != NULL;) {
+        assert(exec->vm_data.valid);
+        _PyExecutorObject *next = exec->vm_data.links.next;
+
+        if (!exec->vm_data.was_run) {
+            unlink_executor(exec);
+            if (PyList_Append(invalidate, (PyObject *)exec) < 0)
+            {
+                goto error;
+            }
```
Comment (picnixz) [inline Python/optimizer.c]:
```suggestion
            if (PyList_Append(invalidate, (PyObject *)exec) < 0) {
                goto error;
            }
```

File: Python/optimizer.c [hunk 15]
```diff
@@ -1657,6 +1667,49 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
...
+
+        } else {
```
Comment (picnixz) [inline Python/optimizer.c]:
```suggestion
        }
        else {
```

File: Include/internal/pycore_optimizer.h [hunk 16]
```diff
@@ -31,7 +31,8 @@ typedef struct {
     uint8_t oparg;
     uint16_t valid:1;
     uint16_t linked:1;
-    uint16_t chain_depth:14;  // Must be big engough for MAX_CHAIN_DEPTH - 1.
+    bool was_run:1;
```
Comment (picnixz) [inline Include/internal/pycore_optimizer.h]: While it shouldn't change on modern compilers, I think compilers like consecutive types, so I'd suggest using a `uint16_t:1` (I think the structure will have `sizeof` 64, whether you use bool or uint16_t, at least on gcc, but I'm not sure on other compilers; it would also be consistent with the types of `valid` and `linked`).

File: Include/internal/pycore_optimizer.h [hunk 17]
```diff
@@ -123,11 +124,16 @@ PyAPI_FUNC(PyObject *) _PyOptimizer_NewUOpOptimizer(void);
...
+PyAPI_FUNC(void) _Py_Executors_InvalidateCold(PyInterpreterState *interp);
+
 #else
 #  define _Py_Executors_InvalidateDependency(A, B, C) ((void)0)
 #  define _Py_Executors_InvalidateAll(A, B) ((void)0)
+#  define _Py_Executors_InvalidateCold(A) ((void)0)
+
 #endif
 
+#define JIT_CLEANUP_THRESHOLD 10
```
Comment (picnixz) [inline Include/internal/pycore_optimizer.h:139]:
```suggestion
// Maximum number of executors that can be kept alive.
#define JIT_CLEANUP_THRESHOLD 10
```

Or something like this ("alive" is maybe not the right word so use the best one you can think of).

File: Include/internal/pycore_interp.h [hunk 18]
```diff
@@ -259,7 +259,7 @@ struct _is {
     struct callable_cache callable_cache;
     _PyOptimizerObject *optimizer;
     _PyExecutorObject *executor_list_head;
-
+    size_t executors_created;
```
Comment (picnixz) [inline Include/internal/pycore_interp.h]: I'd suggest adding a comment saying that this number is limited to `JIT_CLEANUP_THRESHOLD`

File: Python/optimizer.c [hunk 19]
```diff
@@ -182,6 +182,12 @@ _PyOptimizer_Optimize(
     if (err <= 0) {
         return err;
     }
+
+    if (++interp->executors_created >= JIT_CLEANUP_THRESHOLD) {
+        interp->executors_created = 0;
```
Comment (picnixz) [inline Python/optimizer.c]: You could move `interp->executors_created = 0` into `_Py_Executors_InvalidateCold`. That way, calling `_Py_Executors_InvalidateCold` does not require you do pre-update this value.

The reason I'm suggesting this is that `_Py_Executors_InvalidateAll` is already modifying `interp->executors_created`, so I think `_Py_Executors_InvalidateCold` should also be responsible for this.

File: Python/optimizer.c [hunk 20]
```diff
@@ -1659,4 +1669,42 @@ _Py_Executors_InvalidateAll(PyInterpreterState *interp, int is_invalidation)
...
+    PyObject *invalidate = PyList_New(0);
+    if (invalidate == NULL) {
+        goto error;
+    }
+
+    /* Clearing an executor can deallocate others, so we need to make a list of
+     * executors to invalidate first */
+    for (_PyExecutorObject *exec = interp->executor_list_head; exec != NULL;) {
+        assert(exec->vm_data.valid);
+        _PyExecutorObject *next = exec->vm_data.links.next;
```
Comment (picnixz) [inline Python/optimizer.c:1686]: Isn't it possible to have a cycle if don't unlink it in this loop?

EDIT: nvm, the construction doesn't seem to allow it.

File: Include/internal/pycore_optimizer.h [hunk 21]
```diff
@@ -31,7 +31,8 @@ typedef struct {
     uint8_t oparg;
     uint16_t valid:1;
     uint16_t linked:1;
-    uint16_t chain_depth:14;  // Must be big engough for MAX_CHAIN_DEPTH - 1.
+    uint16_t was_run:1;
```
Comment (markshannon) [inline Include/internal/pycore_optimizer.h]: Using a single bit means that setting it will take a read, mask and write instead of a just a write
Maybe reorganize this as:
```
    uint8_t valid:1;
    uint8_t linked:1;
    uint8_t chain_depth:6; // Must be big enough for MAX_CHAIN_DEPTH - 1.
    bool was_run;
```

File: Python/optimizer.c [hunk 22]
```diff
@@ -182,6 +182,12 @@ _PyOptimizer_Optimize(
     if (err <= 0) {
         return err;
     }
+
+    if (++interp->executors_created >= JIT_CLEANUP_THRESHOLD) {
```
Comment (markshannon) [inline Python/optimizer.c]: This could behave quite badly with many executors.

Suppose we have a 1000 executors:
Even in the ideal case, we need to scan all 1000, every time we create 10 new ones.

But what worries me is that we could clear all executors far too easily.
Suppose the VM starts compiling some new region of code, resulting in 20 new executors.
After the first 10, we will mark all executors as cold. After the second 10 we will clear all executors.

I'm not sure what the solution is, but I think a ticker that is independent of the number of executors created would be better. Maybe the GC trigger, or something similar?

**Review by markshannon — COMMENTED** [review]
Looks promising. I have a couple of comments.

Do we have any stats or performance numbers on this?

File: Python/optimizer.c (hunk 22 above)
Comment (savannahostrowski) [reply Python/optimizer.c]: In earlier runs of the benchmarks, there was [no impact to performance](https://github.com/python/cpython/pull/123402#issuecomment-2319188505), even in instances where I tried pretty aggressive thresholds. In discussing this with Brandt, it seemed cheap to re-create in cases where we might be overzealous in invalidation.

**Comment by savannahostrowski** [comment]
@markshannon Thanks for taking a look. The preliminary results looked something like this: https://github.com/python/cpython/pull/123402#issuecomment-2319188505. The approach has been updated since then, so it's worth running the benchmarks again. @brandtbucher, what are your thoughts about re-running the benchmarks at this point?

File: Python/optimizer.c (hunk 22 above)
Comment (savannahostrowski) [reply Python/optimizer.c]: Probably also worth mentioning that I did experiment with triggering invalidation on gc earlier but in chatting with Brandt, it seemed like better design to have this tied to executor creation (i.e. since the invalidation isn't really at all coupled to gc today).

I'm sure that @pablogsal probably has thoughts here too?

Comment (pablogsal) [reply Python/optimizer.c]: > I'm sure that @pablogsal probably has thoughts here too?

I think it may be sensible to do it in the eval breaker or at any other similar point in the VM execution where we can know it's safe to do. This is equivalent to the GC (as both are triggered there) and still allows us to add custom triggering logic and decuples the run from the GC.

For context: I would still prefer not to tie this with the gc as in general we have been holding off coupling the gc with things for a long time because any bug in the GC or on GC objects is already a very non trivial and hard to debug spooky-action-at-a-distance and we have learn that (1) controlling any side-effects that are not purely derived from the generic algorithm and (2) keeping the GC from having to know what's operating on has been key to keep complexity at bay.

Comment (brandtbucher) [reply Python/optimizer.c]: > I'm not sure what the solution is, but I think a ticker that is independent of the number of executors created would be better. Maybe the GC trigger, or something similar?

I think the current approach is probably good enough for now, and we can iterate on the heuristics. The potential quadratic behavior (and wipe-out-everything-ness) of the current approach is a bit scary, but is unlikely to seriously impact most programs. It can also be mitigated by a follow-up PR to change "every ten executors" to "10% more executors" or "20% more JIT memory" or some other ratio-based approach, like the GC uses.

I also think the GC is probably the wrong place for this. We *could* add a new dedicated eval-breaker bit, but that sort of feels like overkill. The creation of new executors is a known safe point where (a) we *expect* to spend a bit of time managing JIT stuff and (b) we know JIT memory pressure will increase by some fixed amount. I don't think it would be any better to just set an eval breaker bit here and leave our new JIT code almost immediately to perform this same cleanup.

**Comment by brandtbucher** [comment]
> @brandtbucher, what are your thoughts about re-running the benchmarks at this point?

I don't believe the main heuristic has changed, and the little improvements in bit-twiddling and such are unlikely to impact performance or memory significantly. We probably don't need to re-run anything.

File: Include/internal/pycore_interp.h [hunk 23]
```diff
@@ -259,7 +259,8 @@ struct _is {
     struct callable_cache callable_cache;
     _PyOptimizerObject *optimizer;
     _PyExecutorObject *executor_list_head;
-
+    // executors_created is limited to JIT_CLEANUP_THRESHOLD
+    size_t executors_created;
```
Comment (brandtbucher) [inline Include/internal/pycore_interp.h]: I'd rename this to `new_executors` or something similarly descriptive. The current name makes it sound like a total count, rather than "executors created since the last cleanup".

File: Include/internal/pycore_optimizer.h (hunk 21 above)
Comment (brandtbucher) [reply Include/internal/pycore_optimizer.h]: Good idea. And just to bikeshed the name, I like a name like `warm` a bit better than `was_run`. `was_run` makes me think that this indicates whether the executor has *ever* been run, which isn't true.

File: Python/bytecodes.c [hunk 24]
```diff
@@ -4840,6 +4840,10 @@ dummy_func(
             assert(((_PyExecutorObject *)executor)->vm_data.valid);
         }
 
+        tier2 op(_SET_EXECUTOR_RUN_STATE, (--)) {
```
Comment (brandtbucher) [inline Python/bytecodes.c]: Not a big deal, but I find the name a tiny bit wordy and not super descriptive. Maybe `_MAKE_WARM` or `_WARM_UP` or something?

**Review by brandtbucher — APPROVED** [review]
Just a couple notes on naming, otherwise looks like a clear improvement to me. Thanks for your patience!

**Comment by savannahostrowski** [comment]
Succeeded by #124443
